### DISCUSSION TITLE
[Feature] External library upload

### DISCUSSION THREAD

**Discussion by peterdstone** [immich-app/immich#5453] [Feature Request]
### The feature

I've gone for a workflow where 2 of us share an account and use syncthing to move files to server, exiftool to copy to destination and immich to serve images.

the request is 3 parts.
1. for an option to have a library at admin level for all or selected users
2. allow upload to an external library or shared location.
3. raised elsewhere... delete from library.

I tried immich upload with exif tool but got duplicate entries, half broken.

excellent work btw. clean interface but most importantly fast and workable with 7Mb upload.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by treardon17** [comment]
My use case for this:

I take a bunch of photos on my DSLR while I'm on a trip, and I want to store the RAW photos in a `year/trip_name/raw` folder, and put all my edits in `year/trip_name/edits`. Each day I back up my photos to the `raw` folder. I want to be able to manage (upload/move/delete) all the photos on a per-trip basis, and it would be very convenient to manage all of this within the same application. I used to do this with google drive/google photos when they were connected, and then switched to synology photos, but would prefer an open source solution, and completely switch to immich.

**Reply by aisbergde** [reply]
if you used synology photos, then you own a Synology.

My current workaround is
- using Synology Drive to upload photos
- mounting this folder as external library
- enable the immich automatic scan of new content
- using external software like digiKam to manage my assets

It is only a workaround. Using immich upload directly would be better and easier.

**Reply by treardon17** [reply]
@aisbergde that makes sense. I think my main issue is that I want a single place to manage and consume my photos. I'm not sure I really see the point of me using Immich if it's only a way to consume the photos. The photo management is just as useful if not more useful to me than the ability to view the photos.

I will hold off using Immich until there is a folder view, and either the ability to define a custom folder structure in an internal library or the ability to upload/delete files in an external library. Until then, Synology photos works well enough. It's just resource constrained on a lower powered machine and is not open source.

**Reply by atomicpumpkins** [reply]
Folder view is coming in the next version. You can automate the moving workflow on the OS level of your choice.
https://github.com/immich-app/immich/discussions/5409#discussioncomment-10390471

**Comment by mjforan** [comment]
+1 for "allow upload to an external library or shared location."
I like to keep all my media organized in the external library. I also like to keep my phone synced using the mobile app or to occasionally upload to the web interface. Manually moving files from the Immich library to my external folder causes problems. I wish I could set uploads to go to a folder in my external library.

**Comment by ldivl** [comment]
I use another storage device to host my media over SMB. It would be great to have the option to upload directly to an external library. Additionally, having the ability to define the folder structure during uploads would be highly beneficial. For example, I currently use the Photosync app, which organizes files in the following structure: basefolder > yyyy > mm.  My workflow works pretty dang good, but it would be better to upload directly so system can perform indexing immediately without having to do scheduled scans or file monitoring.

**Comment by canton7** [comment]
As much as I hate adding "me too" comments, there's another aspect which I don't think has been mentioned.

I like to take photos on my phone and edit on a computer. I've got Owncloud set up to sync photos between my phone, computer and server, and Immich running on the server.

If I upload photos to Immich from my phone, I don't get to edit them on my computer (as you can only upload photos to Immich's library, and you can't change those files from under Immich). If I set them up as an external library in Immich then I can edit them on my computer, but don't get to upload them to Immich from my phone.

(Granted I can upload them to Owncloud from my phone, then use my computer to organise them into folders, create albums on Immich, etc, but it's much more laborious).

Immich is *so close* to supporting this case perfectly: it "just" needs to be able to upload to external libraries from a phone.

Either way, I'm loving Immich. Thank you for the great work ❤️

**Reply by aisbergde** [reply]
I would say: upload to external libraries is one solution. Another solution would be: allow edit assets in external libraries.

**Reply by canton7** [reply]
Does that solve my use-case? I'm editing from a computer (not the web interface), and I can already edit assists in external libraries (edits from my computer are synced back into the external folder using Owncloud).

My hang-up is that getting assets into the external folder from my phone, and creating albums for them in Immich, is a bit long-winded, when it's so close to being really smooth.

**Reply by aisbergde** [reply]
maybe I didn't understand your issue.
I am using Synology and Synology Drive to upload, and as soon as it is on my NAS, it is available via external libraries in immich. I don't understand, what could be the issue here. 
My issue: limited possibilities on assets in external libraries (edit, ...)

**Reply by canton7** [reply]
It's as I describe. The Immich mobile app is wonderful for letting you select a bunch of images easily, and assigning them to a new album there and then. It's quick and convenient, no computer needed, no need to log into the Immich web interface to create a new album, etc. The only downside is that there's no way to do this to external storage, so it doesn't work with my set-up.

Instead, I have to let Owncloud sync the assets, then go in on my computer and manually sort the assets into folders, create an album on Immich, assign the assets to an album, etc. It works, but it's much less convenient than just a couple of taps from the mobile app.

**Reply by aisbergde** [reply]
maybe I didn't understand your issue.
I am using Synology and Synology Drive to upload, and as soon as it is on my NAS, it is available via external libraries in immich. I don't understand, what could be the issue here. 
My issue: limited possibilities on assets in external libraries (edit, ...)

**Reply by bo0tzz** [reply]
@canton7 I don't know what editing tools you use, but it seems to me like your case would be best solved with some sort of integration from your editor (as a plugin or such) to interact with assets in Immich.

**Reply by canton7** [reply]
It would be solved perfectly by the request made by others in this discussion: to allow uploads to external libraries. 

IMO that's far more likely to happen (and would amost certainly work better) than reworking a photo editor to support using a third-party web-based photo management app.

**Comment by bdovaz** [comment]
One more vote... https://github.com/immich-app/immich/discussions/16892

**Comment by madwhofr** [comment]
Would also love to see this feature be implemented

**Comment by ruocchio** [comment]
+1 for this, or something like it.

I wonder if a slightly different approach would work - is it possible to just allow a user to configure their library to a different mount point?  Of course that would be dependent on whether or not a users library react to FS changes dynamically, or is it triggered by an upload/app user action?  Just curious, as this would make all libraries external in a sense, with one being the default.

**Comment by antoniovj1** [comment]
I'd prefer for photos I upload through Immich to be stored in an external folder mounted within Docker. This would keep everything organized and easily accessible through other programs. Thanks!!

**Comment by aandrewjoyce** [comment]
I would also like to tag on to this. I want my database to run on my photo server, but reference an external NAS drive that is redundant, backed up, etc. I think for now it's simpler to not use the immich app on my phone, and use the NAS app instead. Then, Immich can just read the external library and react to updates from the NAS app.

**Comment by antoniovj1** [comment]
This can help:

https://github.com/immich-app/immich/discussions/22442

**Comment by industral** [comment]
I wonder is it so difficult to allow to change path to external library instead of pre-hardcoded one? 
Also, AFAIK it's a good decision to keep your photos on some general dataset and allow apps to use it instead of keeping photos inside some apps datastore.

<img width="1686" height="986" alt="Screenshot 2025-12-29 at 20 21 00" src="https://github.com/user-attachments/assets/f7193365-af96-40c6-a641-d5a029acd547" />

**Comment by thisisgonnatakeawhile** [comment]
Seconding this request (but also for Web). I'd be great if Immich could provide an option to upload media directly to a chosen/selectable External Library (and automatically scan the library once the upload is finished).

This feature could be complimented by providing an option to move/migrate media from an External Library to Immich's UPLOAD_LOCATION without losing any metadata added to the media (and viceversa).

Thank you Immich devs, you guys rock!

**Comment by joveha** [comment]
This is the only feature I am missing in Immich

**Comment by Mannshoch** [comment]
As I read this here. In my opinion many cases would be solved If we instead of using external libraries could use a virtual directory:
https://github.com/immich-app/immich/discussions/7488#discussioncomment-11887340

**Comment by Joshsg02** [comment]
+1 Uploading backups from the phone to multiple external libraries would be great to separate my family's photos from my own. It would also solve the issue of Immich asking users to backup outside of the Immich folder.

This request has been asked by a lot of people and there is no solution on the roadmap yet. Hopefully this gets more attention as I feel this is quite important.

**Comment by tomuta** [comment]
This is the one thing that keeps me from using immich on my NAS. I have a "Pictures" share with only pictures. But there is no way to configure Immich to not only put meta data like thumbnails, databases  etc in a different place (docker volume). I don't want it to store all that in my Pictures share. I can sort of do it by adding my Pictures share as an external library, but then I run into this limitation where I can't configure an upload folder into an external library. I would absolutely LOVE to have this feature because it seems like that would allow me to just have a docker volume for all the cached immich data, databases etc, and then I could add my Pictures share as an external library and when people take pictures on their phone they would simply be added directly to my Pictures share. Right now this keeps me from actually using Immich.

**Comment by snowfiend131** [comment]
Adding a +1 to this topic.

I need the ability to manage photos outside of Immich to support my work flow, including moving photos from the phone camera upload folder into a broader photo directory that I manage with Lightroom, which includes photos from a real non-phone camera.  This need pushes me towards an External Library, given the limitations of not being able to touch photos in the Internal Library outside of Immich.

However, use of an External Library ruins the use of Immich as a day-to-day real-time photo app on my phone.  It takes too long for a 3rd party upload tool to upload the photo to an external library folder, then for the Immich database to sync and recognize this photo and show in the External Library.  Meaning, I can't take a photo and immediately see it in the Immich app.  Then I have to use a different app phone app for my real-time photo viewing needs.

So one of two changes are needed to support my needs:

1.  Open up the Internal Library limitations, to let you mess with or remove photos inside it, without breaking the database.  Regularly scan the internal library for missing assets and remove from the database.

or 

2. Allow Immich to upload directly to an External Library, and immediately import this photo into Immich.  Then, I could take a photo and immediately see it in the Immich app.

**Comment by Reisu83** [comment]
I installed Immich on Proxmox a few days ago and added my QNAP as External library, where I have 1,5 TB of pictures. I have so far used QuMagie to backup my pictures as well as my wifes and childrens to the QNAP to individual folders. The problem with QNAP is that it needs to authenticate very often and my family doesn't check and do this.

The problem with Immich, as already said here, is that all users pictures are uploaded to the same folder, which everyone has access to. This is not optimal, as I especially my teenaged kids take a lot of silly pictures, they don't want me or their siblings to look at.

One solution is to put up 5 x LXC containers with Immich spinning in parallel, but I'm not sure this is the best solution if you don't have endless of resources.

What I'm asking for is to be able to choose the upload location per user, but ideally you should also be able to give the log in credentials for each location. Optionally I could create a Immich user, which I grant access to all upload floders and mount the folders using those credentials.

**Reply by bo0tzz** [reply]
> The problem with Immich, as already said here, is that all users pictures are uploaded to the same folder, which everyone has access to.

Idk what you're doing exactly but in standard Immich this is definitely not the case.

**Comment by Puntjuh** [comment]
Adding a +1 as well. 

I now manage my photo's through TrueNAS, it serves as my SMB share for photo's for which I use Nextcloud on my phone and my wife's phone to upload photo's to the specified share. My wife and i each have our own share that we upload to and we each have our own IMMICH account with our own share set as the external drive, which IMMICH reads from. However, i would like to cut out the middle man and simply use IMMICH to upload photos to the external share. 

The reason i don't want to use IMMICH's own storage/database is that I do nightly RSYNC jobs from my TrueNAS to pCloud and another storage server i have, so that i have external backups.

**Reply by bo0tzz** [reply]
> The reason i don't want to use IMMICH's own storage/database is that I do nightly RSYNC jobs from my TrueNAS to pCloud and another storage server i have, so that i have external backups.

I don't see how these relate exactly? You can back up Immich's internal storage perfectly fine as well. (also nitpick, but rsync is not a backup tool)
